LookAndFeels/foleys_Skeuomorphic.cpp

```diff
@@ -0,0 +1,241 @@
+    g.fillEllipse (bounds);
+}
+
+const Skeuomorphic::knobImages& Skeuomorphic::getKnobImages (int diameter)
+{
+    auto search = knobsBgs.find (diameter);
+    if (search != knobsBgs.end()) return search->second;
+
+    float radius = diameter * 0.5f;
+    bool isTiny  = diameter < 20;
+    bool isSmall = diameter < 60;
+
+    juce::Point<float>     centre { radius, radius };
+    juce::Rectangle<float> bounds { 0.0f, 0.0f, float (diameter), float (diameter) };
+
+    if (knobsBgs.size() > size_t (knobsMaxSize))
+        knobsBgs.clear();
+
+
+    // background
+    juce::Image bg { juce::Image::ARGB, diameter, diameter + 3, true };
+    {
+        juce::Graphics g (bg);
+        fillEllipse (g, bounds.translated (0.0f, 1.0f), whiteA010);
+        fillEllipse (g, bounds.translated (0.0f, 2.0f), whiteA010);
+        fillEllipse (g, bounds,                         { 50, 51, 61});
+
+        auto backgroundGr = isTiny? makeGradient (centre, { centre.x, centre.y + radius }, true,
+                                               {  { 0.0, juce::Colours::transparentBlack },
+                                                  { 0.9, juce::Colours::transparentBlack },
+                                                  { 1.0, blackA122 }  }) :
+                                    makeGradient (centre, { centre.x, centre.y + radius }, true,
+                                               {  { 0.0,                          juce::Colours::transparentBlack },
+                                                  { 0.52,                         blackA092 },
+                                                  { 0.9,                          juce::Colours::transparentBlack },
+                                                  { double (1 - (4.0f / radius)), juce::Colours::transparentBlack },
+                                                  { 1.0,                          blackA092 } });
+        fillEllipse (g, bounds, backgroundGr);
+    }
+
+
+    bounds.reduce (diameter * 0.15f, diameter * 0.15f);
+    auto fgRadius = bounds.getWidth() * 0.5f;
+
+
+    // foreground
+    juce::Image fg { juce::Image::ARGB, diameter, diameter, true };
+    {
+        juce::Graphics g (fg); // lower shadow
+        {
+            auto xOffset = fgRadius * 0.12f;
+            auto yOffset = fgRadius * 0.32f;
+
+            juce::ColourGradient shadowGr = makeGradient ({ centre.x + xOffset, centre.y + yOffset },
+                                                          { centre.x, centre.y + fgRadius + yOffset }, true,
+                                                       {  { 0.0,  blackA122                 },
+                                                          { 0.72, blackA122                 },
+                                                          { 1.0,  juce::Colours::transparentBlack }  });
+            fillEllipse (g, bounds.translated (xOffset, yOffset), shadowGr);
+        }
+
+        // middle
+        {
+            g.saveState();
```

Comment:
Please use `juce::Graphics::ScopedSaveState save (g);` instead

Reply:
I will need to keep that code inside another set of curly brackets for this to work, like:
```
        // middle
        {
            {
                juce::Graphics::ScopedSaveState save (g);
                clipToEllipse (g, bounds);

                juce::ColourGradient gr { { 55, 55, 66 }, 0.0f, bounds.getY(),
                                          { 34, 35, 41 }, 0.0f, bounds.getBottom(), false };
                fillEllipse (g, bounds, gr);

                juce::ColourGradient ringGr = makeGradient (centre, { centre.x, centre.y + fgRadius }, true,
                                                         {  { 0.0,  juce::Colours::transparentBlack },
                                                            { 0.72, juce::Colours::transparentBlack },
                                                            { 1.0,  blackA042                 }  });
                fillEllipse (g, bounds, ringGr);

                g.setColour (whiteA042);
                g.drawEllipse (bounds.translated (0.0f, 1.0f), 1.0f);
                if (!isSmall) g.drawEllipse (bounds.translated (0.0f, 2.0f), 1.0f);
            }

            g.setColour (blackA092);
            g.drawEllipse (bounds, 1.0f);
        }
```

Is that ok?

Reply:
I'm assuming it is, going ahead :)